Resolve `DependencyResolver` issue with missing cyclic sub-graphs

loganharbour opened the pull request:
Closes #26575

loganharbour committed:
Add lots of useful assertions to TheWarehouse

loganharbour committed:
Simplify cache construction

loganharbour committed:
Check for non-visited nodes

refs #26575

loganharbour committed:
Add reasonable assertion for sorted size

refs #26575

loganharbour committed:
Add a test case that would have previously not thrown

refs #26575

loganharbour committed:
Avoid O(N) search, keep looking for cyclic dep

refs #26575

loganharbour commented:
I'm still going to add in the extra assertions to `TheWarehouse` that got me to this point

loganharbour committed:
Make sure the resolver gives us all objects back

refs #26575

loganharbour committed:
Make sure this object is always added, add assertion

lindsayad commented on framework/include/interfaces/DependencyResolverInterface.h:
```diff
@@ -82,35 +82,33 @@ DependencyResolverInterface::sortDFS(typename std::vector<T> & vector)
   for (auto & v : vector)
   {
+    // Whether or not this object supplies something, we will always
+    // add it as node because we want to make sure that it gets returned
```
```suggestion
    // add it as a node because we want to make sure that it gets returned
```

lindsayad commented on framework/include/utils/DependencyResolver.h:
```diff
@@ -278,7 +282,24 @@ DependencyResolver<T, Compare>::dfs()
+  {
+    // At this point, if we have any sub-graphs that are cyclic that do not have any
+    // roots _and_ we have found one more or more separate sub-graphs with a root,
+    // we will have never visited the afformentioned cyclic sub-graphs. Therefore,
```
```suggestion
    // we will have never visited the aforementioned cyclic sub-graphs. Therefore,
```

lindsayad commented on framework/include/utils/DependencyResolver.h:
```diff
@@ -405,7 +429,12 @@ bool
 {
   bool cyclic = false;
-  _visited[root] = true;
+  auto & visited = _visited.at(root);
```
```suggestion
  auto & visited = libmesh_map_find(_visited, root);
```

loganharbour committed:
Addres review: make words great again

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

loganharbour committed:
Address review: use better map find

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

loganharbour commented:
@GiudGiud take a look at your failing positions test: https://civet.inl.gov/job/2005939/

In CLI args you end up adding a cyclic dependency that ends up expanding like this

```
[Positions]
  [input]
    type = InputPositions
    positions = '0 1 0
                 1 0 2'
    outputs = none
    initial_positions = file # cyclic
  []
  [file]
    type = FilePositions
    files = '../../multiapps/positions_from_file/positions.txt'
    outputs = none
    initial_positions = input # cyclic
  []
  ...
```

When I remove `initial_positions` from both, it still passes. Ideas?

lindsayad commented on framework/include/interfaces/DependencyResolverInterface.h:
```diff
@@ -122,7 +122,7 @@ DependencyResolverInterface::cyclicDependencyError(CyclicDependencyException<T2>
   std::vector<std::string> names(cycle.size());
   for (const auto i : index_range(cycle))
-    names[i] = static_cast<T>(cycle[i])->name();
+    names[i] = static_cast<T>(cycle[i])->typeAndName();
```
I kind of liked how concise the output was before. Seems like this could be a lot of text for a large cycle. I also don't know why there is a trailing slash after the name in `typeAndName()`

loganharbour replied:
There isn't a trailing slash? It's `ObjectType "name"`:

```
  /**
   * Get the class's combined type and name; useful in error handling.
   * @return The type and name of this class in the form '<type()> "<name()>"'.
   */
  std::string typeAndName() const
  {
    return type() + std::string(" \"") + name() + std::string("\"");
  }
```

It made it a lot easier for me to debug, but I'm not invested. Will remove

lindsayad replied:
Lol, read the string wrong. If this was really helpful for you, then let's keep it

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2006048/) on 54208f8 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26576/site)

This comment will be updated on new commits.

GiudGiud commented:
I think this test was showing that when querying positions at initial, the initial_positions was not queried at initial. (so no infinite series of calls if the intiial position of the initial position was the first position). 
If that s leading to a cycle in the object creation dependency resolution, feel free to modify the test

loganharbour committed:
Base the assertion on unique objects

loganharbour committed:
Remove cyclic dependency

This test still tests using initial positions, it just
doesn't do it for both objects.

loganharbour commented:
> I think this test was showing that when querying positions at initial, the initial_positions was not queried at initial. (so no infinite series of calls if the intiial position of the initial position was the first position). If that s leading to a cycle in the object creation dependency resolution, feel free to modify the test

Got it.

I removed the cyclic dependency, but the test is still able to test initial positions for a single object instead of both.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2006740/) on 0a4ea3c wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>61ed12</th>
      <th colspan=4>#26576 0a4ea3</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.08%</td>
      <td>85.08%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>98650</td>
      <td>98658</td>
      <td>+8</td>
      <td>22</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17297</td>
      <td>17295</td>
      <td>-2</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26576/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26576/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26576/coverage/xfem)

</details>



This comment will be updated on new commits.
